applications/CoSimulationApplication/python_scripts/base_classes/co_simulation_predictor.py

```diff
@@ -66,3 +67,32 @@ def _GetDefaultParameters(cls):
             "data_name"  : "UNSPECIFIED",
             "echo_level" : 0
         }""")
+
+
+class HistoricalDataAccessor:
+    def __init__(self, interface_data, buffer_size):
+        self.interface_data = interface_data
+        self.buffer_size = buffer_size
+        self.step = 0
+
+        self.additional_buffer_size = max(0, buffer_size - interface_data.GetBufferSize())
+        self.aux_data = deque(maxlen=self.additional_buffer_size)
+
+    def CloneTimeStep(self):
+        self.step += 1
+        if self.additional_buffer_size > 0:
+            self.aux_data.appendleft(self.interface_data.GetData())
+
+    def GetData(self, buffer_index):
+        if buffer_index >= self.buffer_size:
+            raise Exception("the buffer-size is not large enough current buffer size: {} | requested solution_step_index: {}!".format(self.buffer_size, buffer_index+1))
+
+        if self.interface_data.GetBufferSize() > buffer_index:
+            # the interface data buffer size is large enough
+            return self.interface_data.GetData(buffer_index)
+        else:
+            # TODO check if time buffer is initialized
```

Comment:
you mean `self.aux_data` ??

Reply:
yes, I forgot to implement it :)

Reply:
So are you implement it in this PR or a separate one ?